<a href="https://colab.research.google.com/github/SultanDaffaN/Data-Mining-Class/blob/main/Tutorial/1.%20EDA/tutorial_logistic_regression_analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Contoh Melakukan Logistic Regression Analysis

By Alfan F. Wicaksono
(Fakultas Ilmu Komputer, UI)

<br/>

In [ ]:
# membuat dataset dummy

import pandas as pd

data = {'weight': [50, 60, 90, 40, 60, 74, 61, 95, 85, 95, 50, 56, 60, 40, 55, 90, 70, 60, 89, 85, 96],
        'age': [20, 23, 24, 25, 27, 28, 30, 40, 45, 48, 49, 49, 49, 50, 57, 57, 60, 67, 68, 64, 70],
        'heart_disease': [0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1]
        }

df = pd.DataFrame(data)

print (df)

    weight  age  heart_disease
0       50   20              0
1       60   23              0
2       90   24              1
3       40   25              0
4       60   27              0
5       74   28              1
6       61   30              0
7       95   40              0
8       85   45              1
9       95   48              0
10      50   49              1
11      56   49              1
12      60   49              0
13      40   50              1
14      55   57              0
15      90   57              1
16      70   60              1
17      60   67              1
18      89   68              1
19      85   64              1
20      96   70              1


In [ ]:
# kita ingin mengetahui effect sizes dari weight dan age terhadap
# heart_disease. Kali ini kita bisa gunakan logistic regression
# analysis

import statsmodels.formula.api as smf

model = smf.logit("""
    heart_disease ~
          weight +
          age
""", data = df)

results = model.fit()
results.summary()

Optimization terminated successfully.
         Current function value: 0.522285
         Iterations 6


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:          heart_disease   No. Observations:                   21
Model:                          Logit   Df Residuals:                       18
Method:                           MLE   Df Model:                            2
Date:                Mon, 13 Feb 2023   Pseudo R-squ.:                  0.2352
Time:                        22:38:19   Log-Likelihood:                -10.968
converged:                       True   LL-Null:                       -14.341
Covariance Type:            nonrobust   LLR p-value:                   0.03428
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -4.2622      2.490     -1.712      0.087      -9.143       0.618
weight         0.0170      0.029      0.585      0.558      -0.040       0.074
age            0.0767      0.038      2.039      0.041       0.003       0.150
==============================================================================
"""

Yang perlu dilihat adalah coefficient, dan p-value (P>|z|). Jika p-value < 0.05, kita katakan bahwa variable tersebut adalah faktor untuk permasalahan binary classification tersebut (karena mempunyai evidence kuat bahwa nilai coefficient-nya berbeda dari nol). Coeff menggambangkan effect size-nya.

Dari hasil analysis dengan data dummy di atas, dapat dilihat bahwa weight bukan merupakan faktor (p-value = 0.558 > 0.05). Namun, age merupakan fitur yang baik karena mempunyai efek terhadap heart_disease (p-value = 0.041 < 0.05).

Nilai coefficient dari age adalah 0.077, yang berarti "jika yang lain (weight) konstan, ketika age bertambah 1 tahun, odds ratio terkena sakit jantung menjadi meningkat e^0.077 = 1.08 kali lipat dari sebelumnya".